src/conv.rs

```diff
@@ -201,61 +192,127 @@ pub fn map_adapter_options<'a>(
 }
 
 pub fn map_device_descriptor<'a>(
-    _: &native::WGPUDeviceDescriptor,
+    des: &native::WGPUDeviceDescriptor,
     extras: Option<&native::WGPUDeviceExtras>,
 ) -> (wgt::DeviceDescriptor<Label<'a>>, Option<String>) {
+    let native_limits = unsafe { (*des.requiredLimits).limits };
+    let mut features = wgt::Features::empty();
+    let limits = map_limits(native_limits);
     if let Some(extras) = extras {
-        let mut features = wgt::Features::empty();
+        // Handle native features speficied in extras
         if (extras.nativeFeatures
             & native::WGPUNativeFeature_TEXTURE_ADAPTER_SPECIFIC_FORMAT_FEATURES)
             > 0
         {
             features |= wgt::Features::TEXTURE_ADAPTER_SPECIFIC_FORMAT_FEATURES;
         }
-        let mut limits = wgt::Limits::default();
-        if extras.maxTextureDimension1D != 0 {
-            limits.max_texture_dimension_1d = extras.maxTextureDimension1D;
-        }
-        if extras.maxTextureDimension2D != 0 {
-            limits.max_texture_dimension_1d = extras.maxTextureDimension2D;
-        }
-        if extras.maxTextureDimension3D != 0 {
-            limits.max_texture_dimension_2d = extras.maxTextureDimension3D;
-        }
-        if extras.maxTextureArrayLayers != 0 {
-            limits.max_texture_array_layers = extras.maxTextureArrayLayers;
-        }
-        if extras.maxBindGroups != 0 {
-            limits.max_bind_groups = extras.maxBindGroups;
-        }
-        if extras.maxDynamicStorageBuffersPerPipelineLayout != 0 {
-            limits.max_dynamic_storage_buffers_per_pipeline_layout =
-                extras.maxDynamicStorageBuffersPerPipelineLayout;
-        }
-        if extras.maxStorageBuffersPerShaderStage != 0 {
-            limits.max_storage_buffers_per_shader_stage = extras.maxStorageBuffersPerShaderStage;
-        }
-        if extras.maxStorageBufferBindingSize != 0 {
-            limits.max_storage_buffer_binding_size = extras.maxStorageBufferBindingSize;
-        }
-        (
+        return (
             wgt::DeviceDescriptor {
                 label: OwnedLabel::new(extras.label).into_cow(),
                 features: features,
                 limits,
             },
             OwnedLabel::new(extras.tracePath).into_inner(),
-        )
+        );
     } else {
-        (
+        return (
             wgt::DeviceDescriptor {
                 label: None,
-                features: wgt::Features::empty(),
+                features,
                 limits: wgt::Limits::default(),
             },
             None,
-        )
+        );
+    }
+}
+
+pub fn map_limits(limits: native::WGPULimits) -> wgt::Limits {
+    let mut wgt_limits = wgt::Limits::default();
+    if limits.maxTextureDimension1D != 0 {
```

Comment:
mmm, I guess this should test against `WGPU_LIMIT_U32_UNDEFINED` instead of 0.

Reply:
Although ... I just realized this will potentially break a lot of people's code in a surprising way. I can't help feeling slightly uncomfortable about the `WGPU_LIMIT_U32_UNDEFINED`. For many limits a zero makes no sense, making it a suitable value for "default". @kvark what do you think?

Reply:
I wrote the code, but still feel uncomfortable. I tried to put it in words in https://github.com/webgpu-native/webgpu-headers/issues/111

Reply:
Thanks! That makes sense to me

Reply:
I propose that we merge this as-is, and once it the API for specifying undefined is final, we change that for all structs in one PR (possibly a separate release too).